torchrl/data/utils.py

```diff
@@ -35,6 +43,206 @@
+            if re.match(r"Found more than one unique shape in the tensors", str(err)):
+                pass
+            else:
+                raise err
+
+    if isinstance(td, TensorDict) and not len(td.keys()):
+        return None
+    return td
+
+
+def _empty_like(td, batch_size):
+    if is_tensor_collection(td):
+        return td.empty()
+    else:
+        shape = [dim if i < len(batch_size) else 0 for i, dim in enumerate(td.shape)]
+
+        return torch.empty(
+            shape,
+            dtype=td.dtype,
+            device=td.device,
+        )
+
+
+def check_no_lazy_keys(td, recurse: bool = True):
+    """Given a TensorDictBase, returns true if there are no lazy keys."""
+    if isinstance(td, LazyStackedTensorDict):
+        keys = set(td.keys())
+        for t in td.tensordicts:
+            if recurse and not check_no_lazy_keys(t):
+                return False
+            if set(t.keys()) != keys:
+                return False
+    elif isinstance(td, TensorDict) and recurse:
+        for i in td.values():
+            if not check_no_lazy_keys(i):
+                return False
+    elif isinstance(td, torch.Tensor):
+        return True
+    else:
+        return False
+
+    return True
+
+
+def get_all_keys(td, include_lazy: bool):
+    """Given a TensorDictBase, returns all lazy and not lazy keys as a set tuples."""
+    keys = set()
+    if isinstance(td, LazyStackedTensorDict) and include_lazy:
+        for t in td.tensordicts:
+            keys = keys.union(get_all_keys(t, include_lazy=include_lazy))
+    if isinstance(td, TensorDictBase):
+        for key in td.keys():
+            try:
+                keys.add((key,))
+                value = td.get(key)
+                inner_keys = get_all_keys(value, include_lazy=include_lazy)
+                for inner_key in inner_keys:
+                    keys.add((key,) + _unravel_key_to_tuple(inner_key))
+            except RuntimeError:
+                pass
+    return keys
+
+
+def all_eq(
+    td: Union[TensorDictBase, torch.Tensor],
+    other: Union[TensorDictBase, torch.Tensor],
+):
+    """Returns true if the two classes match all entries in the keys and stack dimensions."""
+    if td.__class__ != other.__class__:
+        return False
+
+    if td.shape != other.shape or td.device != other.device:
+        return False
+
+    if isinstance(td, LazyStackedTensorDict):
+        if td.stack_dim != other.stack_dim:
+            return False
+        for t, o in zip(td.tensordicts, other.tensordicts):
+            if not all_eq(t, o):
+                return False
+    elif isinstance(td, TensorDictBase):
+        td_keys = list(td.keys())
+        other_keys = list(other.keys())
+        if td_keys != other_keys:
+            return False
+        for k in td_keys:
+            if not all_eq(td[k], other[k]):
+                return False
+    elif isinstance(td, torch.Tensor):
+        return torch.equal(td, other)
+    else:
+        raise AssertionError()
```

Comment:
let's put an explicit message (also why do we have an exception here?)
Can we use something else than AssertionError?
It's basically the same as `assert` which we try to avoid in the code base